dspace-server-webapp/src/main/java/org/dspace/app/rest/ItemIdentifierController.java

```diff
@@ -0,0 +1,166 @@
+ * advanced to queue for reservation or registration.
+ *
+ * @author Kim Shepherd
+ */
+@RestController
+@RequestMapping("/api/" + ItemRest.CATEGORY + "/" + ItemRest.PLURAL_NAME + REGEX_REQUESTMAPPING_IDENTIFIER_AS_UUID
+        + "/" + IdentifiersRest.NAME)
+public class ItemIdentifierController {
+
+    @Autowired
+    ConverterService converter;
+
+    @Autowired
+    ItemService itemService;
+
+    @Autowired
+    ItemRestRepository itemRestRepository;
+
+    @Autowired
+    MetadataConverter metadataConverter;
+
+    @Autowired
+    IdentifierService identifierService;
+
+    @Autowired
+    DOIService doiService;
+
+    @Autowired
+    Utils utils;
+
+    /**
+     * Request that an identifier of a given type is 'created' for an item. Depending on the identifier type
+     * this could mean minting, registration, reservation, queuing for registration later, etc.
+     *
+     * @return 201 CREATED on success, or 302 FOUND if already created, or an error
+     */
+    @RequestMapping(method = RequestMethod.POST)
+    @PreAuthorize("hasPermission(#uuid, 'ITEM', 'ADMIN')")
+    public ResponseEntity<RepresentationModel<?>> registerIdentifierForItem(@PathVariable UUID uuid,
+                                                                            HttpServletRequest request,
+                                                                            HttpServletResponse response,
+                                                                            @RequestParam(name = "type") String type)
+            throws SQLException, AuthorizeException {
+        Context context = ContextUtil.obtainContext(request);
+
+        Item item = itemService.find(context, uuid);
+
+
+        if (item == null) {
+            throw new ResourceNotFoundException("Could not find item with id " + uuid);
+        }
+
+        // Check for a valid identifier type and register the appropriate type of identifier
+        if ("doi".equals(type)) {
+            return registerDOI(context, item);
+        } else {
+            throw new IllegalArgumentException("Valid identifier type (eg. 'doi') is required, parameter name 'type'");
+        }
+    }
+
+    /**
+     * Queue a new, pending or minted DOI for registration for a given item. Requires administrative privilege.
+     * This request is sent from the Register DOI button (configurable) on the item status page.
+     *
+     * @return 302 FOUND if the DOI is already registered or reserved, 201 CREATED if queued for registration
+     */
+    private ResponseEntity<RepresentationModel<?>> registerDOI(Context context, Item item)
+            throws SQLException, AuthorizeException {
+        String identifier = null;
+        HttpStatus httpStatus = HttpStatus.BAD_REQUEST;
+        ItemRest itemRest;
+        try {
+            DOIIdentifierProvider doiIdentifierProvider = DSpaceServicesFactory.getInstance().getServiceManager()
```

Comment:
could doiIdentifierProvider be moved as field of this class? If so we could set it during class initialization, possibly raising an exception if no DOI provider is configured

Reply:
Note: this is now implemented in IdentifierRestRepository.

Wouldn't that cause startup errors for any user that doesn't use DOIs? This REST endpoint needs to be able to start up OK and return data about handles, other identifiers, even if the `createAndReturn` method would throw an error in the case of a missing DOI provider. Does that sound OK?

Reply:
Yes, I see your point and it sounds OK.